### PR TITLE
Add support for reconstruct in IndexShards (#709)

### PR DESCRIPTION
**Pull request by bvarghese1**

### CODE REVIEW

File: MetaIndexes.cpp [hunk 1]
```diff
@@ -564,7 +564,21 @@ void IndexShards::search (
...
-
+void IndexShards::reconstruct (idx_t key, float * recons) const
+{
+    for (int i = 0; i < shard_indexes.size(); i++) {
+        try {
+            shard_indexes[i]->reconstruct (key, recons);
+            if (verbose)
+                printf ("key %ld found in shard %d\n", key, i);
+            return;
+        } catch (const std::out_of_range& e) {
```
Comment (Enet4) [inline MetaIndexes.cpp]: Is it accurate to assume an exception of this type being raised when the key is not found? From what I saw in one of the implementations, [`IndexPQ`](https://github.com/facebookresearch/faiss/blob/master/IndexPQ.cpp#L126), `reconstruct` might throw a `FaissException` instead.

Comment (bvarghese1) [reply MetaIndexes.cpp]: @Enet4 Thanks! Would it useful to catch both the exceptions and have different messages for it in verbose mode or simply replace out_of_range with FaissException for brevity?

**Comment by mdouze** [comment]
Faiss only throws `FaissException`s. 
I reject this PR because we are working on a way to properly propagate exceptions from `IndexShards` sub-indices.

**Comment by bvarghese1** [comment]
@mdouze makes sense. Thanks.

**Comment by mdouze** [comment]
Yes, let's give it another try then.

**Pull request closed without merging**